receiver/dcgmreceiver/scraper.go

```diff
@@ -220,188 +177,180 @@ func discoverRequestedFields(config *Config) []string {
 	return requestedFields
 }
 
-func (s *dcgmScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
-	err := s.initClient()
-	if err != nil || s.client == nil {
-		return s.mb.Emit(), err
+func (s *dcgmScraper) run(ctx context.Context, metricsCh chan<- map[uint]deviceMetrics) error {
+	defer close(metricsCh)
+	for {
+		client, _ := s.initClient()
+		// Ignore the error; it's logged in initClient.
+		if client != nil {
+			s.runOnce(ctx, client, metricsCh)
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case metricsCh <- map[uint]deviceMetrics{}:
```

Comment:
The first usable send/receive is executed, starting at the top of the `select` statement. So if the context is done and `metricsCh` is sendable, it will return and never send on `metricsCh`. (Then the `defer close(metricsCh)` will unblock any receivers, as discussed above.)